### PR TITLE
Migrate matter.js to monorepo

### PR DESCRIPTION
**Pull request by Apollon77**
This is my counter-approach to #51 

it is basically "just" using the workspaces feature of npm and project references of typescript

"nom install" is only called on root level and not anymore in the packges themself. This is important. All rest should work automagically.

What does this PR do:
1.) It creates the packages directory and moves "matter.js" into it; a second dummy package should represend node-matter.content (as Marco did too)
2.) Then tsconfig and build stuff is adjusted to build in this environment
3.) build/test command sin root package.json will execute all sub scripts

Thats basically it

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -1,9 +1,15 @@
-/* Dev configuration */
 {
-  "extends": "./tsconfig.base.json",
-  "compilerOptions": {
-    "esModuleInterop": true, // Required fto remove warnings or jest test
-  },
-  /* During dev, src are included so they are also modified during refactorings. */
-  "include": ["src/**/*.ts", "test/**/*.ts"],
+    "compilerOptions": {
```
Comment (AlCalzone) [inline tsconfig.json:2]: I'd recommend extending from a `@tsconfig/*` package, e.g. https://github.com/tsconfig/bases/#node-lts-tsconfigjson and only add necessary changes.

File: tsconfig.json [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
+        "incremental": true,
```
Comment (AlCalzone) [inline tsconfig.json:3]: composite implies incremental

File: tsconfig.build.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (AlCalzone) [inline tsconfig.build.json:1]: It is a bit cumbersome to manage references everywhere, but I'd recommend basing the `build` config on the non-build config and only make it actually build. E.g. like this.

```jsonc
{
	"extends": "./tsconfig.json",
	// project references here.
	"compilerOptions": {
		"rootDir": "src",
		"outDir": "build"
	},
	"include": ["src/**/*.ts"],
	"exclude": [ /* test files, etc */ ]
}
```

Comment (AlCalzone) [reply tsconfig.build.json:1]: Upside of this, the editor experience is using the same settings, so no surprises at build time

File: tsconfig.build.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        "noImplicitOverride": true,
+        "noUnusedParameters": true,
+        "noUnusedLocals": true,
+        "strictNullChecks": true,
+        "strict": true
+    },
+    "references": [
+        { "path": "packages/matter.js" },
+        { "path": "packages/node-matter" }
+    ]
```
Comment (AlCalzone) [inline tsconfig.build.json:17]: If there's a clear "entrypoint" to the monorepo, one consideration would be to have no references at the top level, and build in the "entrypoint" package instead.

File: packages/matter.js/package.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (mfucci) [inline packages/matter.js/package.json:1]: Can you tell git that this file is from the original package.json and not a new file?
There is a command for that, but I don't remember it.
This will help a lot to see what is different here.

File: packages/node-matter/package.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        "clean": "rm -rf dist",
+        "build": "npm run clean; tsc",
+        "test": "TZ=utc true || mocha"
+    },
+    "devDependencies": {
+        "ts-node": "^10.9.1",
+        "typescript": "^4.9.5"
+    },
+    "dependencies": {
+        "@project-chip/matter.js": "^0.1.0"
```
Comment (mfucci) [inline packages/node-matter/package.json:35]: How does typescript knows to link against the local matter.js build vs the published one?

File: tsconfig.json [hunk 7]
```diff
@@ -1,9 +1,15 @@
...
+        "target": "es2019",
+        "module": "commonjs",
```
Comment (mfucci) [inline tsconfig.json:5]: If this file is the base file for matter.js and node-matter, it should contain only the common parameters: it is more difficult to follow if parameters should be overridden.

File: tsconfig.json [hunk 8]
```diff
@@ -1,9 +1,15 @@
...
+        "declaration": true,
+        "sourceMap": true,
+        "composite": true,
+        "strict": true,
+        "moduleResolution": "node",
+        "esModuleInterop": true,
```
Comment (mfucci) [inline tsconfig.json:11]: This is not be able be default: this are only very specific cases why this is needed that should clearly be listed here. Otherwise, this will create less optimal code...

File: tsconfig.build.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        "noImplicitAny": true,
+        "noImplicitOverride": true,
+        "noUnusedParameters": true,
+        "noUnusedLocals": true,
+        "strictNullChecks": true,
+        "strict": true
+    },
+    "references": [
+        { "path": "packages/matter.js" },
+        { "path": "packages/node-matter" }
```
Comment (mfucci) [inline tsconfig.build.json:16]: packages/node ?

File: packages/node-matter/package.json (hunk 6 above)
Comment (AlCalzone) [reply packages/node-matter/package.json:35]: ~~magic~~ [npm workspaces](https://docs.npmjs.com/cli/v9/using-npm/workspaces)

TL;DR: When a dependency matches the package name of a package referenced by the top-level `package.json` `workspaces`, it automatically gets symlinked instead of installed from elsewhere.

TypeScript just works, it doesn't care where the package comes from btw.

File: tsconfig.json (hunk 8 above)
Comment (AlCalzone) [reply tsconfig.json:11]: Interestingly, this is in the `recommended` base tsconfig:
https://github.com/tsconfig/bases/blob/main/bases/recommended.json

FWIW, the TypeScript team even recommends it in the [release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#support-for-import-d-from-cjs-from-commonjs-modules-with---esmoduleinterop) and imply the recommendation in [the docs](https://www.typescriptlang.org/tsconfig#esModuleInterop):
> Turning on esModuleInterop will fix both of these problems in the code transpiled by TypeScript.

So if anything, this should at least be the default for Node.js

File: packages/node-matter/package.json (hunk 6 above)
Comment (AlCalzone) [reply packages/node-matter/package.json:35]: If you want to try yourself:
```jsonc
// file: package.json
{
	"name": "monorepo",
	"private": true,
	"workspaces": [
		"packages/*"
	]
}
```

```jsonc
// file: packages/a/package.json
{
	"name": "package-a",
	"version": "1.0.0"
}
```

```jsonc
// file: packages/b/package.json
{
	"name": "package-b",
	"version": "1.0.0",
	"dependencies": {
		"package-a": "1.0.0"
	}
}
```

run `npm install` and see how the folders in `node_modules` are actually symlinks to the `packages` subdirectories.

File: packages/matter.js/package.json (hunk 5 above)
Comment (Apollon77) [reply packages/matter.js/package.json:1]: This is the same problem we already had in an other PR: When you move a file AND change it in the same PR then Git will not see it as a "Move" anymore somehow. I did not found a way to prevent that last ime and this time we can not bypass t by using two PRs, lets please accept that behaviour of git/github

File: tsconfig.json (hunk 7 above)
Comment (Apollon77) [reply tsconfig.json:5]: Yes I will clean this up

File: tsconfig.build.json (hunk 9 above)
Comment (Apollon77) [reply tsconfig.build.json:16]: This is the naming discussion we need to do and ideally the directly is names as the package in the end ... The question is how the "new" node-matter package is named ...

Comment (Apollon77) [reply tsconfig.build.json:16]: @turon @mfucci Any ideas? I think Marco had @project-hip/matter-node.js in this PR, we could also leave node-matter ... This is then for the "final" PR

File: packages/matter.js/package.json (hunk 5 above)
Comment (Apollon77) [reply packages/matter.js/package.json:1]: PS: Any yes I remember that you proofed me in another PR that it works if you do that ... but I still did not found out why it is not working when I do that ... no clue and honestly I would love to not spend hours in finding the reason. Maybe my git is too old with 2.19.0 ... need to see how to best update, so will retry in the finalPR ... this is for me just the "proof of concept"

File: tsconfig.json (hunk 1 above)
Comment (Apollon77) [reply tsconfig.json:2]: We could go with @tsconfig/node-lts or with @tsconfig/node16 ... @mfucci @turon What do you think?

File: tsconfig.build.json (hunk 4 above)
Comment (Apollon77) [reply tsconfig.build.json:17]: Because of the fact that we really publish two libraryies and none is the real "entrypoint" I would keep this

File: packages/node-matter/package.json (hunk 6 above)
Comment (Apollon77) [reply packages/node-matter/package.json:35]: Exxactly. This is the magic of the workspaces.

**Comment by Apollon77** [comment]
Ok, I updated and tried to cover the comments best possible ... I especially collected the tsconfigs in the base file ... for the open comments @mfucci @turon need your opinions or ok or whatever.

In my eyes this works and we can optimize configs and such also later.

File: packages/matter.js/package.json (hunk 5 above)
Comment (Apollon77) [reply packages/matter.js/package.json:1]: PS: I assume in this case it is even more special because the "original place" stays also with a package.json with changed content, so it is "moved file and created new in the same spot". maybe would had been registered by git if it would had been two commits? I would not dig into that deeper because feels lost time

File: packages/node-matter/package.json (hunk 6 above)
Comment (Apollon77) [reply packages/node-matter/package.json:35]: Bildschirm­foto 2023-02-24 um 17 58 18

Here it how it looks in the FS in exactly our case

File: tsconfig.build.json (hunk 9 above)
Comment (Apollon77) [reply tsconfig.build.json:5]: Ideally the directories are named like the packages that are inside the directories ... So How the naming should be?

* @project-chip/matter.js we already have

Options for the node package in my eyes
* @project-chip/matter.js-node
* @project-chip/matter-node.js
* @project-chip/matter-nodejs
* ...

Comment (Apollon77) [reply tsconfig.build.json]: Ok we decided for @project-chip/matter-node.js

File: packages/matter.js/package.json (hunk 5 above)
Comment (turon) [reply packages/matter.js/package.json:1]: I've run into the same problem when an existing file is moved away and replaced, git diff just can't figure out how to show the simplest capture of the changes.

File: .gitignore [hunk 10]
```diff
@@ -1,4 +1 @@
-/dist
-/node_modules
-/out
```
Comment (turon) [inline .gitignore:3]: How does git know to ignore `/out`?

File: README.md [hunk 11]
```diff
@@ -1,9 +1,22 @@
-# matter.js
+# JavaScript/TypeScript based Matter Implementation
 
 ![experimental](https://img.shields.io/badge/status-Experimental-red) [![license](https://img.shields.io/badge/license-Apache2-green.svg)](https://raw.githubusercontent.com/project-chip/matter.js/master/LICENSE)
 
 Implementation of Matter protocol in typescript with no native dependencies (and very limited dependencies).
 
+## Monorepo Overview
+
+This repository contains multiple packages (and because of this it is a monorepo). The packages are contained in the `packages` directory and arer all published separately to NPM.
```
Comment (turon) [inline README.md:9]:
```suggestion
This repository contains multiple packages (and because of this it is a monorepo). The packages are contained in the `packages` directory and are all published separately to NPM.
```

File: tsconfig.json (hunk 8 above)
Comment (turon) [reply tsconfig.json:10]: I seem to recall `matternode`, which is mostly `js` as opposed to `ts`, required `esModuleInterop: true` as well to link to `matter.js` properly.

File: README.md [hunk 12]
```diff
@@ -1,9 +1,22 @@
...
+
+* matter.js: the core Matter implementation in typescript which is JavaScript only and has no native dependencies.
+* matter-node.js: a node.js implementation of a Matter
+
+This repository uses the workspaces feature on npm to manage the dependencies between the packages. Because of this please only use `npm install` on the root of the repository. This will install all the dependencies for all the packages and also take care to create relevant symlinks between the packages.
```
Comment (turon) [inline README.md:14]: RE: `only use`, it looks like `npm run build` and `npm test` will work fine from the top-level as well, and simply build or test all workspaces, no?

**Review by turon — APPROVED** [review]
LGTM :+1:

File: .gitignore (hunk 10 above)
Comment (Apollon77) [reply .gitignore:3]: out should not exist on root level anymore becau it only exists in /packages/matter.js/out and it is ignored there

File: README.md (hunk 12 above)
Comment (Apollon77) [reply README.md:14]: yes works too, see in line #18 below.This section is about "npm install" to NOT run npm install in one of the "/packages/* directories

**Review by mfucci — APPROVED** [review]
